src/networking.c

```diff
@@ -4036,8 +4078,9 @@ void flushReplicasOutputBuffers(void) {
          *
          * 3. Obviously if the replica is not ONLINE.
          */
-        if (replica->repl_state == REPLICA_STATE_ONLINE && !(replica->flags & CLIENT_CLOSE_ASAP) &&
-            can_receive_writes && !replica->repl_start_cmd_stream_on_ack && clientHasPendingReplies(replica)) {
+        if ((replica->repl_state == REPLICA_STATE_ONLINE || replica->repl_state == REPLICA_STATE_BG_RDB_LOAD) &&
```

Comment:
These two checks have been repeated several times. I think we are essentially introducing a new conceptual properties for replicas, which is "being ready to accept replication" - can we abstract this notion into a helper function?

```suggestion
        static inline int isReplicaReady(client *replica) {
            return (replica->repl_state == REPLICA_STATE_ONLINE || 
                    replica->repl_state == REPLICA_STATE_BG_RDB_LOAD) && !(replica->flags & CLIENT_CLOSE_ASAP);
        }
        
        if (isReplicaReady(replica) &&
```

Another idea could be reusing/enhancing `canFeedReplicaReplBuffer`?

```c
int canFeedReplicaReplBuffer(client *replica) {
     if (replica->flags & CLIENT_CLOSE_ASAP) return 0;
 
     /* Don't feed replicas that only want the RDB. */
     if (replica->flags & CLIENT_REPL_RDBONLY) return 0;

     /* Don't feed replicas that are still waiting for BGSAVE to start. */
     if (replica->repl_state == REPLICA_STATE_WAIT_BGSAVE_START) return 0;

     return 1;
 }
```

Reply:
I'm afraid that reusing `canFeedReplicaReplBuffer` will let in replicas with unwanted replication states. Lets use a new method for that. WDYT of `isReplicaReadyForReplData`?

Reply:
This is not a blocker to me. We can address this later in a follow up PR.